libvcx/src/api/vcx.rs

```diff
@@ -15,17 +15,117 @@ use crate::utils::error;
 use crate::utils::threadpool::spawn;
 use crate::utils::version_constants;
 
-/// Initializes VCX with config settings
+#[no_mangle]
+pub extern fn vcx_init_threadpool() -> u32 {
+    info!("vcx_init_threadpool >>>");
+    match init_threadpool() {
+        Ok(_) => error::SUCCESS.code_num,
+        Err(err) => error::UNKNOWN_ERROR.code_num
+    }
+}
+
+/// Creates an instance of agency client used to communicate with the agency. Must be called after
+/// wallet was created and opened.
 ///
 /// #Params
 /// command_handle: command handle to map callback to user context.
 ///
-/// config: The agent provision configuration. You can produce this by provisioning agent using function vcx_provision_agent
+/// agency_config: Config of the agency client
+/// {
+///    agency_url - URL of the agency 
+///    agency_did - DID of the agency
+///    agency_pwdid - pairwise DID of the agency created for the connection
+///    agency_vk - verkey  of the agency created for the connection
+///    agent_pwdid - pairwise DID of this client's agent in the agency
+///    agent_vk - verkey of this client's agent in the agency
+///    my_pwdid - pairwise DID of this client used to communicate with it's agent in the agency
+///    my_vk - verkey of this client used to communicate with it's agent in the agency
+/// }
 ///
-/// cb: Callback that provides error status of initialization
+/// cb: Callback that provides error status
+///
+/// #Returns
+/// Error code as a u32
+#[no_mangle]
+pub extern fn vcx_create_agency_client(command_handle: CommandHandle, config: *const c_char, cb: Option<extern fn(xcommand_handle: CommandHandle, err: u32)>) -> u32 {
+    info!("vcx_create_agency_client >>>");
+
+    check_useful_c_str!(config, VcxErrorKind::InvalidOption);
+    check_useful_c_callback!(cb, VcxErrorKind::InvalidOption);
+
+    trace!("vcx_create_agency_client >>> config: {}", config);
+
+    spawn(move || {
+        match init_agency_client(&config) {
+            Ok(()) => {
+                info!("create_agency_client_cb >>> command_handle: {}, rc {}", command_handle, error::SUCCESS.code_num);
+                cb(command_handle, error::SUCCESS.code_num)
+            }
+            Err(e) => {
+                error!("create_agency_client_cb >>> command_handle: {}, error {}", command_handle, e);
+                cb(command_handle, e.into());
+                return Ok(());
+            }
+        }
+        Ok(())
+    });
+    error::SUCCESS.code_num
+}
+
+
+/// Stores institution did, verkey and name in memory.
+///
+/// #Params
+/// command_handle: command handle to map callback to user context.
+///
+/// issuer_config: Pool configuration
+/// {
+///     "institution_name" - Name of the issueing / verifying institution
+///     "institution_did" (optional) - Institution did obtained on vcx_configure_issuer_wallet
```

Comment:
According to implementation this doesn't seem to be optional. If it's not provided, `CONFIG_INSTITUTION_DID` /   `CONFIG_INSTITUTION_VERKEY`  is not set, which would cause runtime error when trying to perform operation relying on these configurations (such as `vcx_issuer_create_credential`)

Reply:
Yes, but it is optional according to implementation of the documented function, and the point of making it optional is that if `vcx_configure_issuer_wallet` was called in the same runtime session before `vcx_init_issuer_config`, `institution_did` and `institution_verkey` does not have to be saved in memory, as this was done in `vcx_configure_issuer_wallet`.

Reply:
I am not necessarily arguing strongly for making them optional, just explaining that the point of making them optional is due to this use case.